File: python/nedb/relations.py
# SPDX-FileCopyrightText: 2026 INTERCHAINED LLC
# SPDX-License-Identifier: BUSL-1.1
# aof-db · a distribution of NEDB · © 2026 INTERCHAINED LLC × Eth-Interchained × Vex (Claude Opus 5)

"""
nedb.relations — first-class, time-travel-aware relations (the graph layer).

Relations are stored as adjacency lists for O(1) traversal. Each edge records the
seq at which it was added and (optionally) removed, so relation queries can also be
asked "AS OF" any past sequence — the graph time-travels just like the records do.
"""
from __future__ import annotations

from typing import Dict, List, Optional, Tuple


class Relations:
    def __init__(self) -> None:
        # (frm, rel) -> list of [to, added_seq, removed_seq|None]
        self._adj: Dict[Tuple[str, str], List[list]] = {}
        # (to, rel) -> list of [frm, added_seq, removed_seq|None]  (reverse index)
        self._radj: Dict[Tuple[str, str], List[list]] = {}

    def link(self, frm: str, rel: str, to: str, seq: int) -> None:
        for e in self._adj.get((frm, rel), []):
            if e[0] == to and e[2] is None:
                return  # already linked
        self._adj.setdefault((frm, rel), []).append([to, seq, None])
        self._radj.setdefault((to, rel), []).append([frm, seq, None])

    def unlink(self, frm: str, rel: str, to: str, seq: int) -> None:
        for e in self._adj.get((frm, rel), []):
            if e[0] == to and e[2] is None:
                e[2] = seq
        for e in self._radj.get((to, rel), []):
            if e[0] == frm and e[2] is None:
                e[2] = seq

    @staticmethod
    def _live(edges, as_of):
        out = []
        for node, added, removed in edges:
            if as_of is None:
                if removed is None:
                    out.append(node)
            else:
                if added <= as_of and (removed is None or removed > as_of):
                    out.append(node)
        return out

    def neighbors(self, frm: str, rel: str, as_of: Optional[int] = None) -> List[str]:
        return self._live(self._adj.get((frm, rel), []), as_of)

    def inbound(self, to: str, rel: str, as_of: Optional[int] = None) -> List[str]:
        return self._live(self._radj.get((to, rel), []), as_of)

File: rust/crates/nedb-core/src/relations.rs
// SPDX-FileCopyrightText: 2026 INTERCHAINED LLC
// SPDX-License-Identifier: BUSL-1.1
// aof-db · a distribution of NEDB · © 2026 INTERCHAINED LLC × Eth-Interchained × Vex (Claude Opus 5)

// nedb-core — first-class, time-travel-aware relations (the graph layer).
//
// Relations are stored as adjacency lists for O(1) traversal. Each edge records
// the seq at which it was added and optionally removed, so traversal can also be
// asked "AS OF" any past sequence — the graph time-travels just like the records do.

use std::collections::HashMap;

/// A single directed edge: target node key, seq added, seq removed (None = live).
#[derive(Clone, Debug)]
pub struct Edge {
    pub to:      String,
    pub added:   u64,
    pub removed: Option<u64>,
}

/// (frm_key, relation) → list of edges
type Adj = HashMap<(String, String), Vec<Edge>>;

#[derive(Default)]
pub struct Relations {
    adj:  Adj,   // forward edges
    radj: Adj,   // reverse edges (to, rel) → [frm, ...]  stored as Edge{to: frm, ...}
}

fn live(edges: &[Edge], as_of: Option<u64>) -> Vec<String> {
    edges
        .iter()
        .filter(|e| match as_of {
            None => e.removed.is_none(),
            Some(s) => e.added <= s && e.removed.map_or(true, |r| r > s),
        })
        .map(|e| e.to.clone())
        .collect()
}

impl Relations {
    pub fn link(&mut self, frm: &str, rel: &str, to: &str, seq: u64) {
        // idempotent: skip if edge already live
        let fwd = self.adj.entry((frm.to_string(), rel.to_string())).or_default();
        if fwd.iter().any(|e| e.to == to && e.removed.is_none()) {
            return;
        }
        fwd.push(Edge { to: to.to_string(), added: seq, removed: None });
        self.radj
            .entry((to.to_string(), rel.to_string()))
            .or_default()
            .push(Edge { to: frm.to_string(), added: seq, removed: None });
    }

    pub fn unlink(&mut self, frm: &str, rel: &str, to: &str, seq: u64) {
        if let Some(fwd) = self.adj.get_mut(&(frm.to_string(), rel.to_string())) {
            for e in fwd.iter_mut() {
                if e.to == to && e.removed.is_none() {
                    e.removed = Some(seq);
                }
            }
        }
        if let Some(rev) = self.radj.get_mut(&(to.to_string(), rel.to_string())) {
            for e in rev.iter_mut() {
                if e.to == frm && e.removed.is_none() {
                    e.removed = Some(seq);
                }
            }
        }
    }

    pub fn neighbors(&self, frm: &str, rel: &str, as_of: Option<u64>) -> Vec<String> {
        self.adj
            .get(&(frm.to_string(), rel.to_string()))
            .map(|edges| live(edges, as_of))
            .unwrap_or_default()
    }

    pub fn inbound(&self, to: &str, rel: &str, as_of: Option<u64>) -> Vec<String> {
        self.radj
            .get(&(to.to_string(), rel.to_string()))
            .map(|edges| live(edges, as_of))
            .unwrap_or_default()
    }

    /// All (frm_key, rel, to_key) triples that are currently live. Used by nedbd to
    /// populate the schema graph without a full Studio deploy.
    pub fn all_live_edges(&self) -> Vec<(String, String, String)> {
        let mut out = Vec::new();
        for ((frm, rel), edges) in &self.adj {
            for e in edges {
                if e.removed.is_none() {
                    out.push((frm.clone(), rel.clone(), e.to.clone()));
                }
            }
        }
        out
    }
}
